noria-server/mir/src/optimize.rs

```diff
@@ -1,19 +1,235 @@
+            MirNodeType::Aggregation { ref on, .. }  => {
+                // if the child is a filter and it has exactly one parent,
+                // then this is a candidate
+                if let MirNodeType::Filter { ref conditions } = child.inner {
+                    if child.ancestors.len() != 1 {
+                        continue;
+                    }
+                    candidate = true;
+
+                    // But wait -- need to check if the filter is on the aggregation result
+                    use nom_sql::FunctionExpression::{Count, Sum};
+                    use nom_sql::FunctionArguments;
+                    for (i, col) in child.columns.iter().enumerate() {
+                        match col.function.as_deref() {
+                             Some(Count(FunctionArguments::Column(ref col), _))
+                             | Some(Sum(FunctionArguments::Column(ref col), _)) if col.name == on.name => {
+                                // this column may be the aggregation result
+                                // so if we're filtering on it, we're not a candidate
+                                for (j, _cond) in conditions {
+                                    if *j == i {
+                                        candidate = false;
+                                        break;
+                                    }
+                                }
+                            },
+                            _ => {},
+                        }
+                    }
+                }
+            },
+            _ => {},
+        };
+        if candidate {
+            candidate_nodes.push(n);
+        }
+    }
+
+    // 3. For each candidate, merge it, and update all parents/children
+    // of the newly merged node.
+
+    let mut new_nodes = Vec::new();
+
+    for n in candidate_nodes {
+        let temp = n.borrow();
+        let c = temp.children.first().unwrap();
+        let child = c.borrow();
+
+        // determine which is the aggregation, and update the conditions if the second isn't filter
+        let (cond, agg) = match n.borrow().inner {
+            MirNodeType::Aggregation { .. }  => {
+                if let MirNodeType::Filter { ref conditions } = child.inner {
+                    // for each column in filter's conditions, we need to figure out
+                    // which column in input to aggregation has the same name so we
+                    // can reindex the condition to that column
+                    let mut conds = Vec::new();
+                    for (i, cond) in conditions {
+                        let mut found = 0;
+                        // conditions are on the parent columns of the filter node, i.e. agg's columns
+                        let cond_col = n.borrow().columns[*i].clone();
+                        // we need to reindex them onto the new filteragg's parent columns, i.e. agg's parent's columns
+                        assert!(n.borrow().ancestors.len() == 1);
```

Comment:
A handy macro for these types of comparisons is `assert_eq!`. It does the same thing, but if the values are _not_ equal, it will print what the values were to help with debugging. There is also `assert_ne!` for the reverse case.